agent/runner/runner.go

```diff
@@ -0,0 +1,252 @@
+	"golang.org/x/sync/semaphore"
+	"google.golang.org/protobuf/types/known/timestamppb"
+
+	"github.com/percona/pmm/agent/runner/actions"
+	"github.com/percona/pmm/agent/runner/jobs"
+	"github.com/percona/pmm/api/agentpb"
+)
+
+const (
+	bufferSize           = 256
+	defaultActionTimeout = 10 * time.Second // default timeout for compatibility with an older server
+	defaultCapacity      = 32
+)
+
+// Runner executes jobs and actions.
+type Runner struct {
+	l *logrus.Entry
+
+	actions chan actions.Action
+	jobs    chan jobs.Job
+
+	actionsResults chan agentpb.AgentRequestPayload
+	jobsMessages   chan agentpb.AgentResponsePayload
+
+	sem *semaphore.Weighted
+	wg  sync.WaitGroup
+
+	rw      sync.RWMutex
+	rCancel map[string]context.CancelFunc
+}
+
+// New creates new runner. If capacity is 0 then default value is used.
+func New(capacity uint16) *Runner {
+	l := logrus.WithField("component", "runner")
+	if capacity == 0 {
+		capacity = defaultCapacity
+	}
+
+	l.Infof("Runner capacity set to %d.", capacity)
+
+	return &Runner{
+		l:              l,
+		actions:        make(chan actions.Action, bufferSize),
+		jobs:           make(chan jobs.Job, bufferSize),
+		sem:            semaphore.NewWeighted(int64(capacity)),
+		rCancel:        make(map[string]context.CancelFunc),
+		jobsMessages:   make(chan agentpb.AgentResponsePayload),
+		actionsResults: make(chan agentpb.AgentRequestPayload),
+	}
+}
+
+// Run starts jobs execution loop. It reads jobs from the channel and starts them in separate goroutines.
+func (r *Runner) Run(ctx context.Context) {
+	for {
+		select {
+		case action := <-r.actions:
+			r.handleAction(ctx, action)
+		case job := <-r.jobs:
+			r.handleJob(ctx, job)
+		case <-ctx.Done():
+			r.wg.Wait() // wait for all actions and jobs termination
+			close(r.actionsResults)
+			close(r.jobsMessages)
+			return
+		}
+	}
+}
+
+// StartAction starts given actions.Action.
+func (r *Runner) StartAction(action actions.Action) error {
+	select {
+	case r.actions <- action:
+		return nil
+	default:
+		return errors.New("actions queue overflowed")
+	}
+}
+
+// StartJob starts given jobs.Job.
+func (r *Runner) StartJob(job jobs.Job) error {
+	select {
+	case r.jobs <- job:
+		return nil
+	default:
+		return errors.New("actions queue overflowed")
+	}
+}
+
+// JobsMessages returns channel with Jobs messages.
+func (r *Runner) JobsMessages() <-chan agentpb.AgentResponsePayload {
+	return r.jobsMessages
+}
+
+// ActionsResults return chanel with Actions results payload.
+func (r *Runner) ActionsResults() <-chan agentpb.AgentRequestPayload {
+	return r.actionsResults
+}
+
+// Stop stops running Action or Job.
+func (r *Runner) Stop(id string) {
+	r.rw.RLock()
+	defer r.rw.RUnlock()
+
+	// Job removes itself from rCancel map. So here we only invoke cancel.
+	if cancel, ok := r.rCancel[id]; ok {
+		cancel()
+	}
+}
+
+// IsRunning returns true if Action or Job with given ID still running.
+func (r *Runner) IsRunning(id string) bool {
```

Comment:
Is the `id` in argument unique across jobs and actions?

Reply:
Yep, they have this format:
`/action/uuid`
`/job/uuid`